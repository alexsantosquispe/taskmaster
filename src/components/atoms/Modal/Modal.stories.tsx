import type { Meta, StoryObj } from '@storybook/react';

import { Wrapper } from '@/utils/testing/unitTest.util';
import { useState } from 'react';
import { Button } from '../Button/Button';
import { Modal } from './Modal';

const meta = {
  title: 'Atoms/Modal',
  component: Modal,
  parameters: {
    layout: 'centered'
  }
} satisfies Meta<typeof Modal>;

export default meta;

type Story = StoryObj<typeof Modal>;

export const Default: Story = () => {
  const [isOpen, setIsOpen] = useState(false);

  return (
    <Wrapper>
      <div className="flex flex-col">
        <Button
          ariaLabel="Button to open modal"
          label="Open modal"
          onClick={() => setIsOpen(true)}
        />
        {isOpen && (
          <Modal title="Modal" onClose={() => setIsOpen(false)}>
            <div>Modal content</div>
          </Modal>
        )}
      </div>
    </Wrapper>
  );
};

Default.args = {};
