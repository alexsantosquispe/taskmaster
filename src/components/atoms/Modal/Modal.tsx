import useKeyPress from '@/hooks/useKeyPress';
import { CloseIcon } from '@/icons';
import type { ReactNode } from 'react';
import { twMerge } from 'tailwind-merge';
import { IconButton } from '../IconButton/IconButton';

interface ModalProps {
  title?: string;
  children: ReactNode;
  onClose: () => void;
  classNames?: {
    container?: string;
    header?: string;
    body?: string;
  };
}

export const Modal = ({ title, onClose, children, classNames }: ModalProps) => {
  useKeyPress('Escape', onClose);

  return (
    <div
      className={
        'bg-primary/70 fixed inset-0 z-50 flex items-end justify-center backdrop-blur-xs md:items-center md:p-8'
      }
    >
      <div
        data-testid="modal"
        className={twMerge(
          'dark:bg-primary flex w-fit flex-col rounded-t-xl bg-white md:min-w-[25rem] md:rounded-xl dark:border dark:border-white/5',
          classNames?.container
        )}
      >
        <div
          className={twMerge(
            'flex items-center justify-between border-b border-neutral-200 px-2 py-1 dark:border-white/20',
            classNames?.header
          )}
        >
          {title && <h3 className="p-2 font-bold">{title}</h3>}
          <IconButton
            ariaLabel="Close modal button"
            onClick={onClose}
            icon={<CloseIcon />}
            tooltipMessage="Close"
            isDefault={false}
          />
        </div>

        <div className={twMerge('flex p-2', classNames?.body)}>{children}</div>
      </div>
    </div>
  );
};
