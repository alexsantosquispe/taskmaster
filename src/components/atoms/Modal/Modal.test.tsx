import { fireEvent, render, screen } from '@testing-library/react';

import { Modal } from './Modal';

describe('Modal', () => {
  const onCloseMock = jest.fn();
  const props = {
    title: 'Modal',
    onClose: onCloseMock
  };

  it('should render the modal', () => {
    const component = render(
      <Modal {...props}>
        <div>Modal content</div>
      </Modal>
    );

    expect(component).toMatchSnapshot();

    expect(screen.getByText('Modal')).toBeInTheDocument();
  });

  describe('behavior', () => {
    beforeEach(() => {
      jest.clearAllMocks();
    });

    it('should close the modal when click the close button', () => {
      render(
        <Modal {...props}>
          <div>Modal content</div>
        </Modal>
      );

      const closeButton = screen.getByRole('button');

      expect(closeButton).toBeInTheDocument();

      fireEvent.click(closeButton);

      expect(onCloseMock).toHaveBeenCalledTimes(1);
    });

    it('should close the modal when pressing Escape key', () => {
      render(
        <Modal {...props}>
          <div>Modal content</div>
        </Modal>
      );

      fireEvent.keyDown(window, { key: 'Escape' });

      expect(onCloseMock).toHaveBeenCalledTimes(1);
    });

    it('should not close the modal when pressing another key', () => {
      render(
        <Modal {...props}>
          <div>Modal content</div>
        </Modal>
      );

      fireEvent.keyDown(window, { key: 'Enter' });

      expect(onCloseMock).not.toHaveBeenCalled();
    });
  });
});
