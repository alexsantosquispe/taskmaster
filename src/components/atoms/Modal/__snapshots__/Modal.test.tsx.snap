// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`Modal should render the modal 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="bg-primary/70 fixed inset-0 z-50 flex items-end justify-center backdrop-blur-xs md:items-center md:p-8"
      >
        <div
          class="dark:bg-primary flex w-fit flex-col rounded-t-xl bg-white md:min-w-[25rem] md:rounded-xl dark:border dark:border-white/5"
          data-testid="modal"
        >
          <div
            class="flex items-center justify-between border-b border-neutral-200 px-2 py-1 dark:border-white/20"
          >
            <h3
              class="p-2 font-bold"
            >
              Modal
            </h3>
            <div
              class="relative flex items-center justify-center"
              data-testid="tooltip-wrapper"
            >
              <button
                aria-label="Close modal button"
                class="text-primary/80 flex w-fit items-center justify-center rounded-lg p-2 transition-colors duration-150 ease-in-out dark:text-white/70 hover:text-primary hover:cursor-pointer hover:bg-neutral-100 dark:hover:bg-white/15 dark:hover:text-white"
                type="button"
              >
                <svg
                  class="size-5"
                  data-testid="close-icon"
                  fill="none"
                  height="24"
                  stroke="currentColor"
                  stroke-linecap="round"
                  stroke-linejoin="round"
                  stroke-width="2"
                  viewBox="0 0 24 24"
                  width="24"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <path
                    d="M18 6 6 18"
                  />
                  <path
                    d="m6 6 12 12"
                  />
                </svg>
              </button>
            </div>
          </div>
          <div
            class="flex p-2"
          >
            <div>
              Modal content
            </div>
          </div>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="bg-primary/70 fixed inset-0 z-50 flex items-end justify-center backdrop-blur-xs md:items-center md:p-8"
    >
      <div
        class="dark:bg-primary flex w-fit flex-col rounded-t-xl bg-white md:min-w-[25rem] md:rounded-xl dark:border dark:border-white/5"
        data-testid="modal"
      >
        <div
          class="flex items-center justify-between border-b border-neutral-200 px-2 py-1 dark:border-white/20"
        >
          <h3
            class="p-2 font-bold"
          >
            Modal
          </h3>
          <div
            class="relative flex items-center justify-center"
            data-testid="tooltip-wrapper"
          >
            <button
              aria-label="Close modal button"
              class="text-primary/80 flex w-fit items-center justify-center rounded-lg p-2 transition-colors duration-150 ease-in-out dark:text-white/70 hover:text-primary hover:cursor-pointer hover:bg-neutral-100 dark:hover:bg-white/15 dark:hover:text-white"
              type="button"
            >
              <svg
                class="size-5"
                data-testid="close-icon"
                fill="none"
                height="24"
                stroke="currentColor"
                stroke-linecap="round"
                stroke-linejoin="round"
                stroke-width="2"
                viewBox="0 0 24 24"
                width="24"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M18 6 6 18"
                />
                <path
                  d="m6 6 12 12"
                />
              </svg>
            </button>
          </div>
        </div>
        <div
          class="flex p-2"
        >
          <div>
            Modal content
          </div>
        </div>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
